### PR TITLE
Add testing instructions to the README

### PR DESCRIPTION
**Pull request by cbezault**
Description
===========
Add instructions for running the tests by hand to the README.

Also fix/skip some flaky tests so I can get a green checkmark.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -220,6 +221,45 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
     msvcp140d_oss.dll
 ```
 
+# How To Run The Tests
+
+In order to run any of the tests you must be in a correctly configured environment for the architecture you are
+targeting, have python 3.8 or newer and have `clang-cl` on the PATH. If you are testing the x64 flavor of the STL then
```
Comment (CaseyCarter) [inline README.md]: We shouldn't assume that every user knows what `clang-cl` is or where it comes from.
```suggestion
targeting, have python 3.8 or newer and have LLVM's `bin` directory on the PATH. If you are testing the x64 flavor of the STL then
```

File: README.md [hunk 2]
```diff
@@ -220,6 +221,45 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
 
+# How To Run The Tests
+
+In order to run any of the tests you must be in a correctly configured environment for the architecture you are
+targeting, have python 3.8 or newer and have `clang-cl` on the PATH. If you are testing the x64 flavor of the STL then
+running from the "x64 Native Tools Command Prompt" is sufficient for setting up the environment.
+
+## Running All The Tests
+
+After configuring the project, running `ctest` from the root of the build output directory will run all the tests.
```
Comment (CaseyCarter) [inline README.md]: Do we really mean "After building the project"? 

Directories don't have roots.
```suggestion
After configuring the project, running `ctest` from the build output directory will run all the tests.
```

File: README.md [hunk 3]
```diff
@@ -220,6 +221,45 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+CTest will only display the standard error output of tests which failed. In order to get more details from CTest's
```
Comment (StephanTLavavej) [inline README.md]: "which" is nonrestrictive, "that" is restrictive, so this should be:

```suggestion
CTest will only display the standard error output of tests that failed. In order to get more details from CTest's
```

File: README.md [hunk 4]
```diff
@@ -220,6 +221,45 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+
+In order to run a subset of one of the testsuites you must invoke `lit` yourself and point it to the desired
+subdirectory of tests.
+
+The CMake configuration step produces a Python script, `${PROJECT_BINARY_DIR}\tests\llvm-lit\llvm-lit.py`.
+`llvm-lit.py` simply calls into the standard `lit` found in `llvm-project\llvm\utils\lit` with certain options
+pre-configured. Namely a mapping of `lit.cfg` files to `lit.site.cfg` files is provided, which informs lit which
+site-specific config file to load whenever it discovers a general config file.
+
+While one can directly use `lit` to run our tests by hand, we only support using the generated `llvm-lit`.
```
Comment (StephanTLavavej) [inline README.md]: I am confused by this entire section. The first paragraph said "you must invoke `lit` yourself", but the third paragraph says that directly using `lit` is unsupported, and all of the examples demonstrate `llvm-lit.py`.

Additionally, users of the test infrastructure shouldn't care about how information is communicated to the underlying `lit`. (This seems like it would be more appropriate as comments in the test infrastructure, or a README in a subdirectory of the test infrastructure.)

Should this section simply explain that `llvm-lit.py` can be invoked with a test suite directory (like `tests\std`), a test category directory (e.g. libcxx's nicely organized subdirectories), or an individual test directory?

File: README.md [hunk 5]
```diff
@@ -220,6 +221,45 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+
+### Examples
+
+```
+C:\Dev\STL\build>cmake -GNinja -DCMAKE_CXX_COMPILER=cl -DCMAKE_TOOLCHAIN_FILE=..\vcpkg\scripts\buildsystems\vcpkg.cmake ..
+
```
Comment (StephanTLavavej) [inline README.md]: Should there be an example of running all of the tests with `ctest`?

File: README.md [hunk 6]
```diff
@@ -220,6 +221,45 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+`llvm-lit.py` simply calls into the standard `lit` found in `llvm-project\llvm\utils\lit` with certain options
+pre-configured. Namely a mapping of `lit.cfg` files to `lit.site.cfg` files is provided, which informs lit which
+site-specific config file to load whenever it discovers a general config file.
+
+While one can directly use `lit` to run our tests by hand, we only support using the generated `llvm-lit`.
+
+### Examples
+
+```
+C:\Dev\STL\build>cmake -GNinja -DCMAKE_CXX_COMPILER=cl -DCMAKE_TOOLCHAIN_FILE=..\vcpkg\scripts\buildsystems\vcpkg.cmake ..
```
Comment (StephanTLavavej) [inline README.md]: Should a comment explain what this line is doing?

File: README.md [hunk 7]
```diff
@@ -220,6 +221,45 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+
+:: This command will run all of the std testsuite.
+C:\Dev\STL\build>python tests\llvm-lit\llvm-lit.py tests\std
+
+:: However if you want to run a subset of std you need to point it to the right place in the sources.
```
Comment (StephanTLavavej) [inline README.md]:
```suggestion
:: However, if you want to run a subset of std, you need to point it to the right place in the sources.
```
Or perhaps this should talk about running all configurations of one test in std?

File: README.md [hunk 8]
```diff
@@ -220,6 +221,45 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+
+While one can directly use `lit` to run our tests by hand, we only support using the generated `llvm-lit`.
+
+### Examples
+
+```
+C:\Dev\STL\build>cmake -GNinja -DCMAKE_CXX_COMPILER=cl -DCMAKE_TOOLCHAIN_FILE=..\vcpkg\scripts\buildsystems\vcpkg.cmake ..
+
+:: This command will run all of the std testsuite.
+C:\Dev\STL\build>python tests\llvm-lit\llvm-lit.py tests\std
```
Comment (StephanTLavavej) [inline README.md]: These commands don't appear to have been tested. This line says `tests\std` but the following line says `..\tests\std`; I suspect that the latter is correct.

File: README.md [hunk 9]
```diff
@@ -220,6 +221,45 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+
+:: However if you want to run a subset of std you need to point it to the right place in the sources.
+C:\Dev\STL\build>python tests\llvm-lit\llvm-lit.py ..\tests\std\tests\VSO_0000000_any_calling_conventions
+
+:: The same applies for libcxx. The following will run all the libcxx containers tests.
+C:\Dev\STL\build>python tests\llvm-lit\llvm-lit.py ..\llvm-project\libcxx\test\std\containers
```
Comment (StephanTLavavej) [inline README.md]: Did we ever tell users to init/update the `llvm-project` submodule? (Or, separately, the `vcpkg` submodule?)

File: README.md [hunk 10]
```diff
@@ -220,6 +221,45 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+subdirectory of tests.
+
+The CMake configuration step produces a Python script, `${PROJECT_BINARY_DIR}\tests\llvm-lit\llvm-lit.py`.
+`llvm-lit.py` simply calls into the standard `lit` found in `llvm-project\llvm\utils\lit` with certain options
+pre-configured. Namely a mapping of `lit.cfg` files to `lit.site.cfg` files is provided, which informs lit which
+site-specific config file to load whenever it discovers a general config file.
+
+While one can directly use `lit` to run our tests by hand, we only support using the generated `llvm-lit`.
+
+### Examples
```
Comment (StephanTLavavej) [inline README.md]: I think should be given equal weight with the preceding subsections:
```suggestion
## Examples
```

File: README.md (hunk 8 above)
Comment (cbezault) [reply README.md]: No both lines work. If you point to `tests\std` there is a `lit.site.cfg` file that gets placed there and will load all the config as necessary for running the entire `std` testsuite. However, if you want to run tests under a specific directory you need to point to the actual source directory.

File: README.md (hunk 4 above)
Comment (cbezault) [reply README.md]: I think I'll add a little bit more context in the examples and see what you think.

File: README.md (hunk 9 above)
Comment (cbezault) [reply README.md]: Something I hadn't considered in my test harness changes is that previously there was no hard requirement to init the llvm-project submodule in order to configure and build the project. Now there is. I should probably go fix that.

File: README.md (hunk 8 above)
Comment (cbezault) [reply README.md]: I'll be consistent and not use examples where the user points to something in the build directory even if they can.

**Comment by BillyONeal** [comment]
#645 should fix the antivirus false positive

File: README.md (hunk 3 above)
Comment (CaseyCarter) [reply README.md]: This doesn't look resolved to me.

Comment (cbezault) [reply README.md]: Thanks for catching that. I usually click resolved as I fix suggestions. Misaligned on this one I guess.

**Comment by cbezault** [comment]
Sorry I force-pushed, something weird was going on.

File: README.md [hunk 11]
```diff
@@ -223,13 +222,24 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
 
 # How To Run The Tests From The Developer Command Prompt For VS
 
-Before running the tests you must configure and build the project as described above. You must also have [Python][] 3.8
-or newer, and have LLVM's `bin` directory on the PATH.
+1. Follow steps 1-9 of
+[How To Build With A Native Tools Command Prompt](#how-to-build-with-a-native-tools-command-prompt).
+2. Invoke `git submodule init llvm-project`
+3. Invoke `git submodule update`
```
Comment (CaseyCarter) [inline README.md]:
```suggestion
2. Invoke `git submodule update --init llvm-project`
```

File: README.md [hunk 12]
```diff
@@ -220,6 +220,64 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+
+:: If you want to run a subset of a testsuite you need to point it to the right place in the sources. The following
+:: will run the single test found under VSO_0000000_any_calling_conventions.
+C:\STL\build>python tests\llvm-lit\llvm-lit.py ..\tests\std\tests\VSO_0000000_any_calling_conventions
+
+:: You can invoke llvm-lit with any arbitrary subdirectory of a testuite. In libcxx this allows you to have a finer
+:: control over what category of tests you would like to run. The following will run all the libcxx map tests.
+C:\STL\build>python tests\llvm-lit\llvm-lit.py ..\llvm-project\libcxx\test\std\containers\associative\map
+```
+
```
Comment (CaseyCarter) [inline README.md]: Should we have another section describing how to interpret the results?

File: README.md [hunk 13]
```diff
@@ -31,9 +31,8 @@ flavor of the STL (native desktop). We need to extend this to build all of the f
 because they need to be updated whenever source files are added/renamed/deleted. We'll delete the legacy machinery as
 soon as possible.)
 
-* Tests: **In progress.** We rely on three test suites: std, tr1, and [libcxx][]. We've partially ported std and fully
-ported libcxx to run under [lit][] using the various configuration/compilers we test internally. tr1 and portions of
-std are still a work in progress.
+* Tests: **In progress.** We rely on three test suites: std, tr1, and [libcxx][]. We've partially ported std and tr1,
```
[the remark this answers is not in the document]
Comment (cbezault) [reply README.md]: The `/clr:pure` and `/BE` tests aren't running.

File: README.md (hunk 12 above)
Comment (cbezault) [reply README.md]: Sure I guess that's worth it since `ctest` has different output than `llvm-lit`.

Comment (CaseyCarter) [reply README.md]: ...and our user may be unfamiliar with both.

File: tests/utils/stl/test/format.py [hunk 14]
```diff
@@ -254,8 +254,11 @@ def getBuildSteps(self, test, lit_config, shared):
 
     def getTestSteps(self, test, lit_config, shared):
         if shared.exec_file is not None:
+            exec_env = test.cxx.compile_env
+            exec_env['TMP'] = str(shared.exec_dir)
```
Comment (CaseyCarter) [inline tests/utils/stl/test/format.py]: Should we also set `TEMP` and `TMPDIR` and `TEMPDIR` for maximum paranoia?

File: README.md [hunk 15]
```diff
@@ -220,6 +220,143 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+
+In the above example we see that twenty-three tests succeeded and five were unsupported.
+
+### Result Code Values
+
+Our tests use the standard [lit result codes][] and the addition of a non-standard result code `SKIP`. For our tests
+one need mostly only concern themselves with the `PASS`, `XFAIL`, `XPASS`, `FAIL`, and `UNSUPPORTED` result codes.
+
+The `PASS` and `FAIL` result codes are self explanatory. We want our tests to `PASS` and not `FAIL`.
+The `XPASS` and `XFAIL` are less obvious. `XFAIL` is actually a successfull result and indicates that we expected the
```
Comment (CaseyCarter) [inline README.md]:
```suggestion
The `XPASS` and `XFAIL` are less obvious. `XFAIL` is actually a successful result and indicates that we expected the

```

File: tests/utils/stl/test/format.py (hunk 14 above)
Comment (cbezault) [reply tests/utils/stl/test/format.py]: The _tempnam documentation states that it only respects `TMP`.

Comment (CaseyCarter) [reply tests/utils/stl/test/format.py]: > This change is pretty targeted to that only.

Yes, I know - I'm asking if you think it's worth widening the scope a bit to preempt similar issues with future tests. I'm not sure *why* I'm asking, since it seems obvious that it would be better to not write such tests in the first place.

Nevermind.

File: README.md [hunk 16]
```diff
@@ -31,8 +31,8 @@ flavor of the STL (native desktop). We need to extend this to build all of the f
 because they need to be updated whenever source files are added/renamed/deleted. We'll delete the legacy machinery as
 soon as possible.)
 
-* Tests: **Coming soon.** We rely on three test suites: std, tr1, and [libcxx][]. We need to replace our current test
-harness, which extensively uses Microsoft-internal machinery.
+* Tests: **In progress.** We rely on three test suites: std, tr1, and [libcxx][]. We've partially ported std and tr1,
+and fully ported libcxx to run under [lit][] using the various configuration/compilers we test internally.
```
Comment (StephanTLavavej) [inline README.md]:
```suggestion
and fully ported libcxx to run under [lit][] using the various configurations/compilers we test internally.
```

File: README.md [hunk 17]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
     msvcp140d_oss.dll
 ```
 
+# How To Run The Tests From The Developer Command Prompt For VS
+
+1. Follow steps 1-9 of
+[How To Build With A Native Tools Command Prompt](#how-to-build-with-a-native-tools-command-prompt).
```
Comment (StephanTLavavej) [inline README.md]: Can these be the cool `[meow][]` form, or does that not work with internal section links?

File: README.md [hunk 18]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+# How To Run The Tests From The Developer Command Prompt For VS
+
+1. Follow steps 1-9 of
+[How To Build With A Native Tools Command Prompt](#how-to-build-with-a-native-tools-command-prompt).
+2. Invoke `git submodule update --init llvm-project`
+3. Invoke `cmake -G Ninja -DCMAKE_TOOLCHAIN_FILE={where your vcpkg clone is located}\scripts\buildsystems\vcpkg.cmake
+-DENABLE_TESTS=TRUE -S . -B {wherever you want binaries}`. This differs from above only in `-DENABLE_TESTS=TRUE`.
+4. If you have already followed the steps from
+[How To Build With A Native Tools Command Prompt](#how-to-build-with-a-native-tools-command-prompt), and have not
+changed the value of {wherever you want binaries} in step four, then there is no need to rebuild to run the tests.
```
Comment (StephanTLavavej) [inline README.md]:
```suggestion
changed the value of `{wherever you want binaries}` in step 4, then there is no need to rebuild to run the tests.
```

File: README.md [hunk 19]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+the tests under that subdirectory. This can mean executing the entirety of a single testsuite, running all tests under
+a category in libcxx, or running a single test in `std` and `tr1`.
+
+## Examples
+
+```
+:: This command configures the project with tests enabled. It assumes you are using the vcpkg submodule and have
+:: already installed boost. It also assumes you have inited and updated the llvm-project submodule.
+C:\STL\build>cmake -GNinja -DCMAKE_CXX_COMPILER=cl -DCMAKE_TOOLCHAIN_FILE=..\vcpkg\scripts\buildsystems\vcpkg.cmake^
+ -DENABLE_TESTS=TRUE ..
```
Comment (StephanTLavavej) [inline README.md]: I think it would be much less confusing to put the space before the caret:
```suggestion
C:\STL\build>cmake -GNinja -DCMAKE_CXX_COMPILER=cl -DCMAKE_TOOLCHAIN_FILE=..\vcpkg\scripts\buildsystems\vcpkg.cmake ^
-DENABLE_TESTS=TRUE ..
```

File: README.md [hunk 20]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+
+:: As stated above this step is only strictly necessary if you have yet to build the STL or if you have changed the
```
Comment (StephanTLavavej) [inline README.md]:
```suggestion
:: As stated above, this step is only strictly necessary if you have yet to build the STL or if you have changed the
```

File: README.md [hunk 21]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+:: output directory of the binaries. Any changes or additions in any of the existing testsuites do not require
+:: recompilation or reconfiguration to take effect when running the tests.
+C:\STL\build>ninja
+
+:: This command will run all of the testsuites with verbose output.
+C:\STL\build>ctest -V
+
+:: This command will run all of the tr1 testuite
```
Comment (StephanTLavavej) [inline README.md]:
```suggestion
:: This command will run all of the tr1 testsuite.
```

File: README.md [hunk 22]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+C:\STL\build>ctest -R tr1
+
+:: This command will also run all of the tr1 testuite.
```
Comment (StephanTLavavej) [inline README.md]:
```suggestion
:: This command will also run all of the tr1 testsuite.
```

File: README.md [hunk 23]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+C:\STL\build>python tests\llvm-lit\llvm-lit.py ..\tests\tr1
+
+:: If you want to run a subset of a testsuite you need to point it to the right place in the sources. The following
+:: will run the single test found under VSO_0000000_any_calling_conventions.
+C:\STL\build>python tests\llvm-lit\llvm-lit.py ..\tests\std\tests\VSO_0000000_any_calling_conventions
+
+:: You can invoke llvm-lit with any arbitrary subdirectory of a testuite. In libcxx this allows you to have a finer
```
Comment (StephanTLavavej) [inline README.md]:
```suggestion
:: You can invoke llvm-lit with any arbitrary subdirectory of a testsuite. In libcxx this allows you to have finer
```

File: README.md [hunk 24]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+:: As stated above this step is only strictly necessary if you have yet to build the STL or if you have changed the
+:: output directory of the binaries. Any changes or additions in any of the existing testsuites do not require
+:: recompilation or reconfiguration to take effect when running the tests.
+C:\STL\build>ninja
+
+:: This command will run all of the testsuites with verbose output.
+C:\STL\build>ctest -V
+
+:: This command will run all of the tr1 testuite
+C:\STL\build>ctest -R tr1
```
Comment (StephanTLavavej) [inline README.md]: Should the two examples with `tr1` instead mention `std` which is the primary test suite that people should be interested in?

File: README.md [hunk 25]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+:: If you want to run a subset of a testsuite you need to point it to the right place in the sources. The following
+:: will run the single test found under VSO_0000000_any_calling_conventions.
+C:\STL\build>python tests\llvm-lit\llvm-lit.py ..\tests\std\tests\VSO_0000000_any_calling_conventions
+
+:: You can invoke llvm-lit with any arbitrary subdirectory of a testuite. In libcxx this allows you to have a finer
+:: control over what category of tests you would like to run. The following will run all the libcxx map tests.
+C:\STL\build>python tests\llvm-lit\llvm-lit.py ..\llvm-project\libcxx\test\std\containers\associative\map
+```
+
+## Interpreting The Results Of Tests
```
Comment (StephanTLavavej) [inline README.md]: It seems confusing that this is the same header level as its "children" `CTest` and `llvm-lit` below.

File: README.md [hunk 26]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+```
+
+CTest will output everything that was sent to stderr for each of the failed testsuites, which can be used to identify
+which individual test within the testsuite failed. It can sometimes be helpful to run CTest with the `-V` option in
+order to see the stdout of the tests.
+
+## llvm-lit
+
+When running the tests directly via the generated `llvm-lit.py` script the result of each test will be printed. The
+format of each result is {[Result Code](#result-code-values)}: {Testsuite Name} :: {Test Name}:{Configuration Number}.
```
Comment (StephanTLavavej) [inline README.md]:
```suggestion
format of each result is `{[Result Code](#result-code-values)}: {Testsuite Name} :: {Test Name}:{Configuration Number}`.
```

File: README.md [hunk 27]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+PASS: tr1 :: tests/cwchar1:25 (26 of 28)
+PASS: tr1 :: tests/cwchar1:26 (27 of 28)
+PASS: tr1 :: tests/cwchar1:27 (28 of 28)
+
+Testing Time: 3.96s
+  Expected Passes    : 23
+  Unsupported Tests  : 5
+```
+
+In the above example we see that twenty-three tests succeeded and five were unsupported.
```
Comment (StephanTLavavej) [inline README.md]:
```suggestion
In the above example we see that 23 tests succeeded and 5 were unsupported.
```

File: README.md [hunk 28]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+
+### Result Code Values
+
+Our tests use the standard [lit result codes][] and the addition of a non-standard result code `SKIP`. For our tests
```
Comment (StephanTLavavej) [inline README.md]: "Our tests use ... the addition" sounds strange. "and a non-standard result code" or "and an additional non-standard result code" would sound better.

File: README.md [hunk 29]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+one need mostly only concern themselves with the `PASS`, `XFAIL`, `XPASS`, `FAIL`, and `UNSUPPORTED` result codes.
```
Comment (StephanTLavavej) [inline README.md]: "one need mostly only concern themselves" is awkward (this document doesn't attempt to avoid "you" elsewhere).

Can this be phrased more simply as "For our tests, only the BLAH result codes are relevant."?

File: README.md [hunk 30]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+
+The `PASS` and `FAIL` result codes are self explanatory. We want our tests to `PASS` and not `FAIL`.
```
Comment (StephanTLavavej) [inline README.md]:
```suggestion
The `PASS` and `FAIL` result codes are self-explanatory. We want our tests to `PASS` and not `FAIL`.
```

File: README.md [hunk 31]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+The `XPASS` and `XFAIL` are less obvious. `XFAIL` is actually a successful result and indicates that we expected the
```
Comment (StephanTLavavej) [inline README.md]:
```suggestion
The `XPASS` and `XFAIL` result codes are less obvious. `XFAIL` is actually a successful result and indicates that we expected the
```
Note that this will overflow 120 columns.

Comment (StephanTLavavej) [inline README.md]: I find it mildly jarring that XPASS and XFAIL are initially mentioned in that order, then explained in reverse order. I would expect XPASS to be explained before XFAIL. (Note that mentioning XFAIL before XPASS would create some inconsistency with the initial explanation of PASS and FAIL.)

File: README.md [hunk 32]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+
```
Comment (StephanTLavavej) [inline README.md]: Spurious empty line.

File: README.md [hunk 33]
```diff
@@ -220,6 +220,144 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+test to fail and it did. `XPASS` is a failure result and indicates that we expected a test to fail but it passed.
+Typically this means that the `expected_results.txt` file for the testsuite needs to be modified and a `FAIL` entry
+needs to be removed.
+
+The `UNSUPPORTED` result code means that the requirements for a test are not met and so it will not be run. Currently
+all tests which use the `/BE` or `clr:pure` flags are unsupported.
```
Comment (StephanTLavavej) [inline README.md]:
```suggestion
all tests which use the `/BE` or `/clr:pure` options are unsupported.
```

File: README.md (hunk 28 above)
Comment (StephanTLavavej) [inline README.md]: The explanation of `SKIP` is skipped.

File: README.md (hunk 17 above)
Comment (cbezault) [reply README.md]: Maybe it does? I'll try it.

File: README.md (hunk 24 above)
Comment (cbezault) [reply README.md]: Sure I was just trying to have at least one example of each. Makes sense though.

File: README.md (hunk 25 above)
Comment (cbezault) [reply README.md]: Woops, the next two were supposed to be nested.

File: README.md (hunk 26 above)
Comment (cbezault) [reply README.md]: I tried that, you can't have links inside of a code snipet.

File: README.md (hunk 17 above)
Comment (cbezault) [reply README.md]: Looks like it does.

File: README.md (hunk 26 above)
Comment (cbezault) [reply README.md]: Maybe I should drop the link?

File: README.md [hunk 34]
```diff
@@ -220,6 +220,150 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+the tests under that subdirectory. This can mean executing the entirety of a single testsuite, running all tests under
+a category in libcxx, or running a single test in `std` and `tr1`.
+
+## Examples
+
+```
+:: This command configures the project with tests enabled. It assumes you are using the vcpkg submodule and have
+:: already installed boost. It also assumes you have inited and updated the llvm-project submodule.
+C:\STL\build>cmake -GNinja -DCMAKE_CXX_COMPILER=cl -DCMAKE_TOOLCHAIN_FILE=..\vcpkg\scripts\buildsystems\vcpkg.cmake ^
+-DENABLE_TESTS=TRUE ..
```
Comment (SuperWig) [inline README.md]: I think these should have a newline between the comments and the commands. It's a bit clustered and hard to parse otherwise.

Comment (cbezault) [reply README.md]: Sure, I can see what it looks like.

File: README.md [hunk 35]
```diff
@@ -341,22 +346,30 @@ Testing Time: 3.96s
...
+those features first the tests will begin passing unexpectedly for us and return `XPASS` results. In order to resolve
+this it is necessary to add a `PASS` entry to the `expected_results.txt` of the testsuite in question.
 
 The `UNSUPPORTED` result code means that the requirements for a test are not met and so it will not be run. Currently
-all tests which use the `/BE` or `clr:pure` flags are unsupported.
+all tests which use the `/BE` or `/clr:pure` options are unsupported.
+
+The `SKIP` result code indicates that a given test was explicitly skipped by adding a `SKIP` entry to the
+`expected_results.txt`. A test may be skipped for a number or reasons, which include, but are not limited to: it being
+an incorrect test; it taking a very long time to run; it failing or passing for the incorrect reason.
```
Comment (StephanTLavavej) [inline README.md]: Fix "number or reasons" typo and change semicolon-delimited list to normal bullet points:
```suggestion
`expected_results.txt`. A test may be skipped for a number of reasons, which include, but are not limited to:
* being an incorrect test
* taking a very long time to run
* failing or passing for the incorrect reason
```

File: README.md [hunk 36]
```diff
@@ -220,6 +220,159 @@ C:\Users\bion\Desktop>dumpbin /IMPORTS .\example.exe | findstr msvcp
...
+### Result Code Values
+
+Our tests use the standard [lit result codes][], and a non-standard result code: `SKIP`. For our tests, only the
+`PASS`, `XFAIL`, `XPASS`, `FAIL`, and `UNSUPPORTED` standard result codes are relevant.
+
+The `PASS` and `FAIL` result codes are self-explanatory. We want our tests to `PASS` and not `FAIL`.
+
+The `XPASS` and `XFAIL` result codes are less obvious. `XPASS` is actually a failure result and indicates that we
+expected a test to fail but it passed. `XFAIL` is a successful result and indicates that we expected the test to fail
+and it did. Typically an `XPASS` result means that the `expected_results.txt` file for the testsuite needs to be
```
Comment (miscco) [inline README.md]: I gues you want a negation here `and it did *not`

Comment (cbezault) [reply README.md]: No I think what is written is correct. `XFAIL` means that we expected the test to fail and it did. `XPASS` means we expected the test to fail but it passed.

I know, it's a little confusing.

Comment (miscco) [reply README.md]: The full sentence is:

>`XFAIL` is a successful result and indicates that we expected the test to fail
and it did.

So we expected the test to fail and it did?
